src/CarbonAware.Aggregators/src/CarbonAware/CarbonAwareAggregator.cs

```diff
@@ -48,41 +48,85 @@ public async Task<IEnumerable<EmissionsForecast>> GetCurrentForecastDataAsync(ID
     {
         using (var activity = Activity.StartActivity())
         {
-            TimeSpan windowSize = GetDurationOrDefault(props);
             _logger.LogInformation("Aggregator getting carbon intensity forecast from data source");
 
             var forecasts = new List<EmissionsForecast>();
             foreach (var location in GetLocationOrThrow(props))
             {
-                var forecast = await this._dataSource.GetCurrentCarbonIntensityForecastAsync(location);
-                var firstDataPoint = forecast.ForecastData.First();
-                var lastDataPoint = forecast.ForecastData.Last();
-                forecast.DataStartAt = GetOffsetOrDefault(props, CarbonAwareConstants.Start, firstDataPoint.Time);
-                forecast.DataEndAt = GetOffsetOrDefault(props, CarbonAwareConstants.End, lastDataPoint.Time + lastDataPoint.Duration);
-                forecast.Validate();
-                forecast.ForecastData = IntervalHelper.FilterByDuration(forecast.ForecastData, forecast.DataStartAt, forecast.DataEndAt);
-                forecast.ForecastData = forecast.ForecastData.RollingAverage(windowSize);
-                forecast.OptimalDataPoint = GetOptimalEmissions(forecast.ForecastData);
-                if (forecast.ForecastData.Any())
-                {
-                    forecast.WindowSize = forecast.ForecastData.First().Duration;
-                }
-                forecasts.Add(forecast);
+                var forecast = await this._dataSource.GetCarbonIntensityForecastAsync(location);
+                var emissionsForecast = ProcessAndValidateForecast(forecast, props);
+                forecasts.Add(emissionsForecast);
             }
 
             return forecasts;
         }
     }
 
+    /// <inheritdoc />
+    public async Task<EmissionsForecast> GetForecastDataAsync(IDictionary props)
+    {
+        using (var activity = Activity.StartActivity())
+        {
+            var locations = GetLocationOrThrow(props);
+            if (locations.Count() > 1)
+            {
+                var ex = new ArgumentException(CarbonAwareConstants.Locations + " field should only contain one location for forecast data.");
```

Comment:
Well this validation is happening at the controller level. So the function that calls this actually only takes a single location and puts it in a list for this function

Reply:
The devil's advocate question here is "If the controller validates this already, then why are we revalidating here?"

But I think the answer to that question is that the WebAPI is not the only interface that we expect to call the aggregator.  We are taking advantage of niceties in the controller (as we should) but we cannot rely on them as our only checks because we expect the CLI to call with props as well.

So I think validating here is the right thing to do, but in general I think that we could improve the user experience in terms of how a user finds out about invalid inputs and how they are directed to resolve them.